spring-integration-core/src/main/java/org/springframework/integration/handler/advice/RateLimiterRequestHandlerAdvice.java

```diff
@@ -0,0 +1,150 @@
+
+	private final RateLimiter rateLimiter;
+
+	private ApplicationEventPublisher applicationEventPublisher;
+
+	/**
+	 * Construct an instance based on default rate limiter options
+	 * and {@value #DEFAULT_NAME} as a rate limiter name.
+	 * @see RateLimiter#ofDefaults
+	 */
+	public RateLimiterRequestHandlerAdvice() {
+		this(RateLimiter.ofDefaults(DEFAULT_NAME));
+	}
+
+	/**
+	 * Construct an instance based on default rate limiter options and provided name.
+	 * @param name the name for the rate limiter.
+	 */
+	public RateLimiterRequestHandlerAdvice(String name) {
+		this(RateLimiter.ofDefaults(name));
+		Assert.hasText(name, "'name' must not be empty");
+	}
+
+	/**
+	 * Construct an instance based on the provided {@link RateLimiter}.
+	 * @param rateLimiter the {@link RateLimiter} to use.
+	 */
+	public RateLimiterRequestHandlerAdvice(RateLimiter rateLimiter) {
+		Assert.notNull(rateLimiter, "'rateLimiter' must not be null");
+		this.rateLimiter = rateLimiter;
+	}
+
+	/**
+	 * Construct an instance based on the provided {@link RateLimiterConfig}
+	 * and {@value #DEFAULT_NAME} as a rate limiter name.
+	 * @param rateLimiterConfig the {@link RateLimiterConfig} to use.
+	 */
+	public RateLimiterRequestHandlerAdvice(RateLimiterConfig rateLimiterConfig) {
+		this(rateLimiterConfig, DEFAULT_NAME);
+	}
+
+	/**
+	 * Construct an instance based on the provided {@link RateLimiterConfig} and name.
+	 * @param rateLimiterConfig the {@link RateLimiterConfig} to use.
+	 * @param name the name for the rate limiter.
+	 */
+	public RateLimiterRequestHandlerAdvice(RateLimiterConfig rateLimiterConfig, String name) {
+		Assert.notNull(rateLimiterConfig, "'rateLimiterConfig' must not be null");
+		Assert.hasText(name, "'name' must not be empty");
+		this.rateLimiter = RateLimiter.of(name, rateLimiterConfig);
+	}
+
+	/**
+	 * Change the {@code limitForPeriod} option of the {@link #rateLimiter}.
+	 * @param limitForPeriod the {@code limitForPeriod} to use.
+	 * @see RateLimiter#changeLimitForPeriod(int)
+	 */
+	public void setLimitForPeriod(int limitForPeriod) {
+		this.rateLimiter.changeLimitForPeriod(limitForPeriod);
+	}
+
+	/**
+	 * Change the {@code timeoutDuration} option of the {@link #rateLimiter}.
+	 * @param timeoutDuration the {@code timeoutDuration} to use.
+	 * @see RateLimiter#changeTimeoutDuration(Duration)
+	 */
+	public void setTimeoutDuration(Duration timeoutDuration) {
+		this.rateLimiter.changeTimeoutDuration(timeoutDuration);
+	}
+
+	/**
+	 * Obtain the metrics from the rate limiter.
+	 * @return the {@link RateLimiter.Metrics} from rate limiter.
+	 * @see RateLimiter#getMetrics()
+	 */
+	public RateLimiter.Metrics getMetrics() {
+		return this.rateLimiter.getMetrics();
+	}
+
+	@Override
+	protected Object doInvoke(ExecutionCallback callback, Object target, Message<?> message) throws Exception {
+		CheckedFunction0<Object> restrictedCall =
+				RateLimiter.decorateCheckedSupplier(this.rateLimiter, callback::execute);
```

Comment:
I thought about that, but then it looks like we can use the same advice instance for different endpoints, so, the `target` and `callback` are going to be fully different from context to context.

Need to sleep with this more, since an internal logic of the advice is shared as well...

Reply:
Hm. Looks like we fully depend on the `invocation.proceed()` which is context-based against a method and actual arguments to call. So, there is no way to decorate only once during initialization. The real call must be decorated any way.

I might miss something, so feel free to share more your thoughts!